src/H5Dmpio.c

```diff
@@ -1816,201 +2102,165 @@ H5D__multi_chunk_filtered_collective_io(H5D_io_info_t *io_info, const H5D_type_i
-
-        /* Set up chunk information for insertion to chunk index */
-        udata.common.layout  = index_info.layout;
-        udata.common.storage = index_info.storage;
-        udata.filter_mask    = 0;
-
-        /* Retrieve the maximum number of chunks being written among all processes */
-        if (MPI_SUCCESS != (mpi_code = MPI_Allreduce(&chunk_list_num_entries, &max_num_chunks, 1,
-                                                     MPI_UNSIGNED_LONG_LONG, MPI_MAX, io_info->comm)))
-            HMPI_GOTO_ERROR(FAIL, "MPI_Allreduce failed", mpi_code)
+        H5D_MPIO_INIT_CHUNK_IDX_INFO(index_info, io_info);
 
-        /* If no one is writing anything at all, end the operation */
-        if (!(max_num_chunks > 0))
-            HGOTO_DONE(SUCCEED);
-
-        /* Allocate arrays for storing MPI file and mem types and whether or not the
-         * types were derived.
+        /* Redistribute any shared chunks being written to and
+         * send around chunk modification data as necessary
          */
-        if (NULL == (file_type_array = (MPI_Datatype *)H5MM_malloc(max_num_chunks * sizeof(MPI_Datatype))))
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate file type array")
-
-        if (NULL == (file_type_is_derived_array = (hbool_t *)H5MM_calloc(max_num_chunks * sizeof(hbool_t))))
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate file type is derived array")
-
-        if (NULL == (mem_type_array = (MPI_Datatype *)H5MM_malloc(max_num_chunks * sizeof(MPI_Datatype))))
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate mem type array")
-
-        if (NULL == (mem_type_is_derived_array = (hbool_t *)H5MM_calloc(max_num_chunks * sizeof(hbool_t))))
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTALLOC, FAIL, "couldn't allocate mem type is derived array")
-
-        /* Iterate over the max number of chunks among all processes, as this process could
-         * have no chunks left to work on, but it still needs to participate in the collective
-         * re-allocation and re-insertion of chunks modified by other processes.
+        if (mpi_size > 1) {
+            chunk_redistribute_func redistribute_func;
+            herr_t                  result;
+
+            /* Determine strategy to employ for redistributing chunks */
+            redistribute_func = H5D__mpio_redistribute_shared_chunks_p0;
+
+            /* Perform redistribution operation */
+            result = (*redistribute_func)(chunk_list, chunk_list_num_entries, io_info, type_info, fm,
+                                          mpi_rank, mpi_size, NULL);
+            if (result < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL, "unable to redistribute shared chunks")
+
+            /* Send chunk modification data */
+            if (H5D__mpio_share_chunk_modification_data(chunk_list, &chunk_list_num_entries, io_info,
```

Comment:
Yes, this is generally where a lot of memory usage will probably be encountered. However, for multi-chunk things get a little bit complicated because a rank can't finish updating a chunk until it's received all messages for that chunk. So we could end up in a situation where ranks are sending messages to other ranks for chunks that those ranks aren't processing yet. Things could be refactored a bit so that everyone keeps a collective list of all the chunks sorted in the same order and then all ranks process in order, either processing the chunk if the rank owns it, or sending a message out if not. That would probably increase memory usage a decent bit though (though probably not as much as the modification message overhead), and would probably have to use blocking communications since chunks are done one at a time. Not the end of the world or anything, but there's certainly room for overlapping non-blocking communications with some parallel compression work.